lib-new/GraphDatabase.coffee

```diff
@@ -0,0 +1,322 @@
+            headers: $(headers).defaults @headers
+            agent: @agent
+            json: body ? true
+            gzip: true  # This is only for responses: decode if gzipped.
+
+        # Important: only pass a callback to Request if a callback was passed
+        # to us. This prevents Request from doing unnecessary JSON parse work
+        # if the caller prefers to stream the response instead of buffer it.
+        , cb and (err, resp) =>
+            if err
+                # TODO: Do we want to wrap or modify native errors?
+                return cb err
+
+            if raw
+                # TODO: Do we want to return our own Response object?
+                return cb null, resp
+
+            if err = Error._fromResponse resp
+                return cb err
+
+            cb null, _transform resp.body
+
+        # Instead of leaking our (third-party) Request instance, make sure to
+        # explicitly return only its internal native ClientRequest instance.
+        # https://github.com/request/request/blob/v2.53.1/request.js#L904
+        # This is only populated when the request is `start`ed, so `start` it!
+        req.start()
+        req.req
+
+    cypher: (opts={}, cb, _tx) ->
+        if typeof opts is 'string'
+            opts = {query: opts}
+
+        if opts instanceof Array
+            opts = {queries: opts}
+
+        {queries, query, params, headers, lean, commit, rollback} = opts
+
+        if not _tx and rollback
+            throw new Error 'Illegal state: rolling back without a transaction!'
+
+        if not _tx?._id and rollback
+            # No query has been made within transaction yet, so this transaction
+            # doesn't even exist yet from Neo4j's POV; nothing to do.
+            cb null, null
+            return
+
+        if commit and rollback
+            throw new Error 'Illegal state: both committing and rolling back!'
+
+        if not _tx and commit is false
+            throw new TypeError 'Can’t refuse to commit without a transaction!
+                To begin a new transaction without committing, call
+                `db.beginTransaction()`, and then call `cypher` on that.'
+
+        if not _tx and not query and not queries
+            throw new TypeError 'Query or queries required'
+
+        if query and queries
+            throw new TypeError 'Can’t supply both a single query
+                and a batch of queries! Do you have a bug in your code?'
+
+        if queries and params
+            throw new TypeError 'When batching multiple queries,
+                params must be supplied with each query, not globally.'
+
+        if queries and lean
+            throw new TypeError 'When batching multiple queries,
+                `lean` must be specified with each query, not globally.'
+
+        method = 'POST'
+        method = 'DELETE' if rollback
+
+        path = '/db/data/transaction'
+        path += "/#{_tx._id}" if _tx?._id
+        path += "/commit" if commit or not _tx
```

Comment:
We should also consider the case when `_tx isnt null and _tx._id is null and commit`. This happens when committing a transaction that has not executed any statements. In Neo4j 2.1.6, the DB considered it a no-op. In 2.1.7 it is an error.

This may also cause errors if `not commit and not rollback and not query and not queries` but I haven't tested that scenario.